model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java

```diff
@@ -121,6 +163,18 @@
      */
     ModelCriteriaBuilder<M> createCriteriaBuilder();
 
+    /**
+     * Returns query parameters builder for current storage.
+     * 
+     * Using {@code QueryParametersBuilder} it is possible to specify various additional parameters for queries
+     * like offset (how many objects to skip before returning the first one), limit (maximum number of objects to 
+     * return) or fields (in form of the corresponding {@link org.keycloak.storage.SearchableModelField}) that will be 
+     * used for ordering resulting streams of objects
+     * 
+     * @return See description. Never return {@code null}
+     */
+    QueryParameters.Builder<M> createQueryParametersBuilder();
```

Comment:
I expected a discussion around this method. I started removing it just before I sent the PR, but then I reverted. What it gives us, to have it this way, is casting to the correct type without any additional work. The `M` template is a model type in `MapStorage`. If we would want to have this directly created in the `*Providers` we would need to provide a model type each time we create it. So, for example, for users, we would need to call this 
```java
QueryParameters.Builder.<UserModel>create()
```

The correct type is necessary so that we have SearchableModelField checked for the correct model.

Reply:
One can get the right type with static methods as well, which could do the trick, see e.g. [Collections.emptySet()](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#emptySet())

Reply:
The problem is caused by the usage of the builder. In some cases we do not put `QueryParameters.Builder` anywhere we just directly create `QueryParameters` from it, and it seems in such cases Java can't infer the type.

Reply:
Could you provide an example?

Reply:
Sure, look at following commit: https://github.com/mhajas/keycloak/commit/66007936937e49d2283b71f20c03fb08743d4b70

The build is failing with:
```
 /home/runner/work/keycloak/keycloak/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:[588,74] incompatible types: org.keycloak.storage.SearchableModelField<org.keycloak.models.UserModel> cannot be converted to org.keycloak.storage.SearchableModelField<java.lang.Object>`
```